python/lsst/ctrl/mpexec/separablePipelineExecutor.py

```diff
@@ -0,0 +1,264 @@
+        builder : `lsst.pipe.base.GraphBuilder`-like, optional
+            A graph builder that implements a
+            `~lsst.pipe.base.GraphBuilder.makeGraph` method. By default, a new
+            instance of `lsst.pipe.base.GraphBuilder` is used.
+
+        Returns
+        -------
+        graph : `lsst.pipe.base.QuantumGraph`
+            The quantum graph for ``pipeline`` as run on the datasets
+            identified by ``where``.
+
+        Notes
+        -----
+        This method does no special handling of empty quantum graphs. If
+        needed, clients can use `len` to test if the returned graph is empty.
+        """
+        if not builder:
+            builder = lsst.pipe.base.GraphBuilder(
+                self._butler.registry,
+                skipExistingIn=self._skip_existing_in,
+                clobberOutputs=self._clobber_output,
+            )
+
+        metadata = {
+            "input": self._butler.collections,
+            "output_run": self._butler.run,
+            "skip_existing_in": self._skip_existing_in,
+            "skip_existing": bool(self._skip_existing_in),
+            "data_query": where,
+            "user": getpass.getuser(),
+            "time": str(datetime.datetime.now()),
+        }
+        graph = builder.makeGraph(
+            pipeline,
+            self._butler.collections,
+            self._butler.run,
+            userQuery=where,
+            metadata=metadata,
+            resolveRefs=True,
+        )
+        _LOG.info(
+            "QuantumGraph contains %d quanta for %d tasks, graph ID: %r",
+            len(graph),
+            len(graph.taskGraph),
+            graph.graphID,
+        )
+        return graph
+
+    def run_pipeline(
+        self,
+        graph: lsst.pipe.base.QuantumGraph,
+        fail_fast: bool = False,
+        graph_executor: Optional[QuantumGraphExecutor] = None,
+    ) -> None:
+        """Run a pipeline in the form of a prepared quantum graph.
+
+        Pre-execution initialization must have already been run;
+        see `pre_execute_qgraph`.
+
+        Parameters
+        ----------
+        graph : `lsst.pipe.base.QuantumGraph`
+            The pipeline and datasets to execute.
+        fail_fast : `bool`, optional
+            If `True`, abort all (parallel) execution if any task fails.
+        graph_executor : `lsst.ctrl.mpexec.QuantumGraphExecutor`, optional
+            A custom graph executor. By default, a new instance of
+            `lsst.ctrl.mpexec.MPGraphExecutor` is used.
+        """
+        if not graph_executor:
+            quantum_executor = SingleQuantumExecutor(
+                self._butler,
+                self._task_factory,
+                skipExistingIn=self._skip_existing_in,
+                clobberOutputs=self._clobber_output,
+            )
+            graph_executor = MPGraphExecutor(
+                numProc=math.ceil(0.8 * multiprocessing.cpu_count()),
+                timeout=86_400.0,
```

Comment:
It'd probably be good to expose these two numbers at a higher level as `__init__` or method arguments.

Reply:
That's something I wanted to specifically avoid doing -- it doesn't seem like something executors in general would need, and it's not obvious that parallelism is worth the effort in PP. Frankly, I find that the heavy use of conditionally required parameters makes the Middleware APIs very hard to use.

(Incidentally, is the default timeout in `CmdLineFwk` really [30 days](https://github.com/lsst/ctrl_mpexec/blob/main/python/lsst/ctrl/mpexec/cmdLineFwk.py#L475)???)

Reply:
Ok, fair enough.  I figured you needed parallelism because you were using `MPGraphExecutor`, would it make sense to just not use that at all?

I think we've basically never seen a failure-to-timeout that we wished _had_ timed out, and plenty of cases where the timeout happened too soon, so we intentionally made the number so high it'd realistically never trigger.

Reply:
So it sounds like timeout, at least, is a formality that's not worth making user-configurable.

As for parallelism, I included it partly because I'm hoping that `SeparablePipelineExecutor` will be useful for others, and partly because I realized while working on this that we'd _never tried_ worker-internal parallelism, which `SimplePipelineExecutor` doesn't support. My comment about whether we need it in PP was more in the context of a future, PP-specific graph executor.